src/graphql_clj/executor.clj

```diff
@@ -208,35 +210,38 @@
                                    (execute-definition context schema resolver-fn definition fragments variables))
                                  operation-definitions))})))
 
+;; Public API
+
+(defn prepare* [schema-or-state resolver-fn ^String statement]
+  (let [state (if (or (:errors schema-or-state) (:spec-map schema-or-state))
+                schema-or-state
+                (validator/validate-schema schema-or-state))] ;; Schema validation inside execution phase for backwards compatibility
+    (if (:errors state)
+      (select-keys state [:errors])
+      (let [validated-statement (-> statement parser/parse (validator/validate-statement state))]
+        (if (-> validated-statement :state :errors)
+          (select-keys (:state validated-statement) [:errors])
+          (let [resolver (resolver/create-resolver-fn (:schema (:state validated-statement)) resolver-fn)]
+            (assoc-in validated-statement [:state :resolver] resolver)))))))
+
+(def prepare (memoize prepare*))
+
 (defn execute
-  ([context schema resolver-fn ^String statement variables]
-   (let [parsed-document (parser/parse statement)
-         schema-resolver-fn (resolver/create-resolver-fn schema resolver-fn)]
-     (cond
-       (insta/failure? schema) (gerror/throw-error (format "Schema is invalid (%s)." schema))
-       (insta/failure? parsed-document) {:errors [(parser/parse-error parsed-document)]}
-       :else (try
-               (execute-document context schema schema-resolver-fn parsed-document variables)
-               (catch Exception e
-                 (if-let [error (ex-data e)]
-                   {:errors [error]}
-                   (throw e)))))))
-  ([context schema resolver-fn ^String statement]
-   (execute context schema resolver-fn statement nil)))
-
-(comment
-  (execute nil (parser/transform (parser/parse "query {user {id}}")) (graphql-clj.type/create-type-meta-fn graphql-clj.type/demo-schema))
-  (execute nil (parser/transform (parser/parse "query {user {id name}}")) (graphql-clj.type/create-type-meta-fn graphql-clj.type/demo-schema))
-  (execute nil (parser/transform (parser/parse "query {user {id name profilePic {url}}}")) (graphql-clj.type/create-type-meta-fn graphql-clj.type/demo-schema))
-  (execute nil (parser/transform (parser/parse "query {user {id name friends {name}}}")) (graphql-clj.type/create-type-meta-fn graphql-clj.type/demo-schema))
-  (execute nil (parser/transform (parser/parse "query {user { ...userFields friends {name}}} fragment userFields on UserType {id name}")) (graphql-clj.type/create-type-meta-fn graphql-clj.type/demo-schema))
-  (execute nil (parser/transform (parser/parse "{
-  __schema {
-    types {
-      name
-    }
-  }
-}"))
-           (graphql-clj.type/create-type-meta-fn graphql-clj.type/demo-schema))
-  (let [type-schema (type/create-schema graphql-clj.introspection/introspection-schema)]
-    (execute nil type-schema nil "query{__schema{types {name kind}}}")))
+  ([validated-document]
```

Comment:
For public api of execute, we should try to keep it to the bare minimal. User can create they own version of execute if needed. 

It is easier to support only one execute function and maintain backward compatibility in the long run.